quantile/tdigest/tdigest.go

```diff
@@ -0,0 +1,285 @@
+	centroidsPool := opts.CentroidsPool()
+	compression := opts.CompressionFactor()
+	mergedCapacity := mergedCapacity(compression)
+	unmergedCapacity := unmergedCapacity(compression)
+
+	var multiplier int64
+	if precision := opts.QuantilePrecision(); precision != 0 {
+		multiplier = int64(math.Pow10(precision))
+	}
+
+	return &tDigest{
+		compression:      opts.CompressionFactor(),
+		multiplier:       multiplier,
+		centroidsPool:    centroidsPool,
+		merged:           centroidsPool.Get(mergedCapacity),
+		mergedCapacity:   mergedCapacity,
+		unmerged:         centroidsPool.Get(unmergedCapacity),
+		unmergedCapacity: unmergedCapacity,
+		minValue:         positiveInfinity,
+		maxValue:         negativeInfinity,
+	}
+}
+
+func (d *tDigest) Merged() []Centroid {
+	return d.merged
+}
+
+func (d *tDigest) Unmerged() []Centroid {
+	return d.unmerged
+}
+
+func (d *tDigest) Add(value float64) {
+	d.add(value, 1.0)
+}
+
+func (d *tDigest) Quantile(q float64) float64 {
+	// compress the centroids first
+	d.compress()
+
+	var (
+		targetWeight = q * d.mergedWeight
+		curWeight    = 0.0
+		lowerBound   = d.minValue
+		upperBound   float64
+	)
+	for i, c := range d.merged {
+		upperBound = d.upperBound(i)
+		if targetWeight <= curWeight+c.Weight {
+			// The quantile falls within this centroid
+			ratio := (targetWeight - curWeight) / c.Weight
+			quantile := lowerBound + ratio*(upperBound-lowerBound)
+			// If there is a desired precision, we truncate the quantile per the precision requirement
+			if d.multiplier != 0 {
+				quantile = math.Trunc(quantile*float64(d.multiplier)) / float64(d.multiplier)
+			}
+			return quantile
+		}
+		curWeight += c.Weight
+		lowerBound = upperBound
+	}
+
+	// NB(xichen): should never get here unless the centroids array are empty
+	// because the target weight should always be no larger than the total weight
+	return math.NaN()
+}
+
+func (d *tDigest) Merge(tdigest TDigest) {
+	merged := tdigest.Merged()
+	for _, c := range merged {
+		d.add(c.Mean, c.Weight)
+	}
+
+	unmerged := tdigest.Unmerged()
+	for _, c := range unmerged {
+		d.add(c.Mean, c.Weight)
+	}
+}
+
+func (d *tDigest) Close() {
+	if d.closed {
+		return
+	}
+	d.closed = true
+	d.centroidsPool.Put(d.merged)
+	d.centroidsPool.Put(d.unmerged)
+}
+
+func (d *tDigest) Reset() {
+	d.closed = false
+	d.merged = d.centroidsPool.Get(d.mergedCapacity)
+	d.mergedWeight = 0.0
+	d.unmerged = d.centroidsPool.Get(d.unmergedCapacity)
+	d.unmergedWeight = 0.0
+	d.minValue = positiveInfinity
+	d.maxValue = negativeInfinity
+}
+
+// compress merges unmerged centroids and merged centroids in place
+// using unmerged centroids as temporary buffer
+func (d *tDigest) compress() {
+	if len(d.unmerged) == 0 {
+		return
+	}
+
+	sort.Sort(centroidsByMeanAsc(d.unmerged))
+
+	var (
+		totalWeight = d.mergedWeight + d.unmergedWeight
+		currWeight  = 0.0
+		currIndex   = 0.0
+		merged      = d.merged
+		unmerged    = d.unmerged
+		buffer      = d.unmerged[:0]
```

Comment:
It now uses an extra buffer for merging centroids. The code should be more readable now.